### DISCUSSION TITLE
TNoodle in python app

### DISCUSSION THREAD

**Discussion by Radishmouse22** [thewca/tnoodle-lib#24] [Q&A]
I'm making a GUI app in python that is just a basic timer, I've made every bit except for the scrambler which I need TNoodle for. I just need a basic system for requesting scrambles and their SVGs and receiving them back so they can be displayed.

**Comment by gregorbg** [comment]
Hey! To get scrambles from TNoodle that are usable in another app, you need to use the core libraries located at https://github.com/thewca/tnoodle-lib. I will move the discussion there.

**Comment by gregorbg** [comment]
The TNoodle scrambling algorithms are currently implemented in Java. In order to use them from Python, you have two options:

1. Set up an API that internally runs Java, where you can do stuff like `GET yourwebsite.com/scrambles/333?count=5` and it returns a JSON of 5 3x3x3 scrambles. This could be embedded to a local server, so instead you do `GET localhost:8080/foo/bar` without requiring Internet access. **Advantage**: Extremely clean implementation. **Disadvantage**: You'd have to maintain a Java server/infrastructure on top of Python.
2. Play around with [Jython](https://www.jython.org/), which specifically provides Python-to-Java bridge invocations. I have no experience in that arena, but it may provide exactly what you're looking for. **Advantage**: Less heavy infrastructure, direct access to the TNoodle code. **Disadvantage**: Somewhat "hacky" (see below)

In order to make Jython work, you need all the `tnoodle-lib` code in one JAR. Clone this repo, open the project root folder (essentially the thing you just cloned) in a terminal and execute `./gradlew :scrambles:shadowJar` (replace `./gradlew` by `gradlew.bat` on Windows). This will yield a JAR file under `scrambles/build/distributions` that you can point Jython to.

No matter which option you choose, you will need to make yourself familiar with the basic Java infrastructure of how to generate a scramble. https://github.com/thewca/tnoodle-samples and specifically the `java-mvp` folder in there is a great place to get started.

Hope this helps!
- Gregor

**Comment by gregorbg** [comment]
The answer lies in the error message: `Make sure Gradle is running on a JDK, not JRE.`

It seems that you have installed the JRE, not the JDK.
- JRE = Java Runtime Environment. Execute programs built by other people
- JDK = Java Development Kit. Same as JRE plus the functionality to compile programs yourself.

You need the latter in order to build the project with Gradle. Please make sure you have JDK installed and report back if the error persists

**Reply by gregorbg** [reply]
This looks like something may be wrong with the Java version. I will update the embedded Gradle runtime, hopefully this fixes the issue. Hang in there!

**Reply by gregorbg** [reply]
Pull the latest changes and try again please :)

**Comment by Radishmouse22** [comment]


**Reply by gregorbg** [reply]
This is an error related to Java, not Kotlin.
Which Java version are you using?

I cannot reproduce this error on my machine. As a hotfix, you can try editing `buildSrc/src/main/kotlin/configurations/Languages.kt` and replace `sourceCompatibility = JavaVersion.VERSION_1_6` with `VERSION_1_8` instead.

**Comment by Radishmouse22** [comment]
Ive done a bit more research and everything works now, Thanks for your excellent support!

**Reply by campos20** [reply]
Please, consider sharing your code with us. We have a [sample repository](https://github.com/thewca/tnoodle-samples) and we could use some python examples as well, if it's open source.
